tools/ctest_retry.bash

```diff
@@ -1,12 +1,52 @@
 #!/bin/bash
 
-fail=0
-total=$1
-# Change the file path manually to where you want to log fails
+usage() {
+  echo "ctest_retry.bash [options]"
+  echo "Reruns ctest until the test passes or it has run 3 times (by default)."
+  echo "  -r <number> Retry the test up to <number> times [Default: 3]"
+  echo "  -d <path> Execute ctest from the given path [Default: Current working directory]"
+  echo "  -t <testname> Execute only <testname> [Default: execute all tests available in <path> ]"
+  echo "  -h displays this usage summary"
+  exit 0
+}
+
+RETRIES=3
+TESTDIR=.
+SPECIFIC_TEST=""
+
+# Check Options
+while getopts ":hr:d:t:" opt; do
+  case "$opt" in
+    h)
+      usage
+      ;;
+    r)
+      RETRIES=$OPTARG
+      ;;
+    d)
+      TESTDIR=$OPTARG
+      ;;
+    t)
+      SPECIFIC_TEST="-R $OPTARG"
+      ;;
+    \?)
+      echo "Invalid option: -$OPTARG" >&2
+      usage
+      exit 1
+      ;;
+    :)
+      echo "Option -$OPTARG requires an argument." >&2
+      exit 1
+      ;;
+  esac
+done
+
+total=$RETRIES
+cd $TESTDIR
 echo "Retrying Ctest up to " $total " times."
 for ((i=1;i<=total;i++))
   do
-    ctest
+    ctest -V $SPECIFIC_TEST
     result=$?
     if [ "$result" == "0" ]
```

Comment:
Why have this `if`?

Reply:
It's checking to see if the ctest call succeeded, meaning the test passed. If the test passed, then it breaks out of the loop and calls the result a success.

As a result, the test has to fail 3 times before we call it a failure. It only has to pass once to be called a success.

Does that answer your question?

Reply:
Yes it does.